### PR TITLE
feat: Modify Source Code for UTG

### PR DESCRIPTION
**Pull request by khareyash05**

### CODE REVIEW

File: pkg/service/utgen/gen.go [hunk 1]
```diff
@@ -222,9 +222,35 @@ func (g *UnitTestGenerator) Start(ctx context.Context) error {
...
+				}
+
+				// modify the source code for refactoring.
+				if strings.Contains(testsDetails.RefactoredSourceCode, "blank output don't refactor code") {
+					if err := os.WriteFile(g.srcPath, []byte(testsDetails.RefactoredSourceCode), 0644); err != nil {
+						return fmt.Errorf("failed to refactor source code:%w", err)
+					}
+					codeModified = true
+					println("sleeping for 5 seconds so that you can see the modified code")
+					time.Sleep(5 * time.Second)
```
Comment (gouravkrosx) [inline pkg/service/utgen/gen.go:244]: Remove this sleep and print statement

File: pkg/service/utgen/gen.go [hunk 2]
```diff
@@ -235,11 +261,32 @@ func (g *UnitTestGenerator) Start(ctx context.Context) error {
...
+				overallCovInc = overallCovInc || coverageInc
+			}
+
+			// if any of the test couldn't increase the coverage, revert the source code
+			if !overallCovInc {
+				err := revertSourceCode(g.srcPath, originalSrcCode, codeModified)
+				if err != nil {
+					utils.LogError(g.logger, err, "Error reverting source code")
+				}
+			} else {
```
Comment (gouravkrosx) [inline pkg/service/utgen/gen.go:286]: No need for else blog, was for debugging.

File: pkg/service/utgen/gen.go [hunk 3]
```diff
@@ -561,6 +622,9 @@ func (g *UnitTestGenerator) getLine(ctx context.Context) (int, error) {
 			utils.LogError(g.logger, err, "Error unmarshalling test line")
 			return 0, err
 		}
+
```
Comment (gouravkrosx) [inline pkg/service/utgen/gen.go:625]: remove this print statement.

File: pkg/service/utgen/gen.go [hunk 4]
```diff
@@ -524,6 +582,9 @@ func (g *UnitTestGenerator) getIndentation(ctx context.Context) (int, error) {
 			utils.LogError(g.logger, err, "Error unmarshalling test headers")
 			return 0, err
 		}
+
+		fmt.Printf("TestDetails: %v\n", testsDetails)
```
Comment (gouravkrosx) [inline pkg/service/utgen/gen.go:586]: remove this print statement.

File: cli/provider/cmd.go [hunk 5]
```diff
@@ -207,7 +207,7 @@ func (c *CmdConfigurator) AddFlags(cmd *cobra.Command) error {
 		cmd.Flags().Int("max-iterations", 5, "The maximum number of iterations.")
 		cmd.Flags().String("test-dir", "", "Path to the test directory.")
 		cmd.Flags().String("llm-base-url", "", "Base URL for the AI model.")
-		cmd.Flags().String("model", "gpt-4o", "Model to use for the AI.")
+		cmd.Flags().String("model", "o1-preview", "Model to use for the AI.")
```
Comment (gouravkrosx) [inline cli/provider/cmd.go:210]: Revert this, because api-server is controlling the model

File: pkg/service/utgen/gen.go [hunk 6]
```diff
@@ -235,11 +259,32 @@ func (g *UnitTestGenerator) Start(ctx context.Context) error {
...
+			}
+
+			// if any of the test couldn't increase the coverage, revert the source code
+			if !overallCovInc {
+				err := revertSourceCode(g.srcPath, originalSrcCode, codeModified)
+				if err != nil {
+					utils.LogError(g.logger, err, "Error reverting source code")
+				}
+			} else {
+				println("Coverage increased, modifying the source code in the prompt builder")
```
Comment (gouravkrosx) [inline pkg/service/utgen/gen.go:285]: remove this else condition also.

**Comment by gouravkrosx** [comment]
@khareyash05 Please address the last comment and also fix the CodeQL

File: pkg/service/utgen/assets/test_generation.toml [hunk 7]
```diff
@@ -65,13 +74,21 @@ The following packages are already installed in the environment. Use these when
 {{ .installed_packages | trim }}
 =========
 
-{{ if .additional_includes_section | trim }}
+
+{%- if additional_includes_section | trim %}
 {{ .additional_includes_section | trim }}
-{{ end }}
+{% endif %}
+
```
Comment (shivamsouravjha) [inline pkg/service/utgen/assets/test_generation.toml:81]: i removed endif 
image
don't put it back

File: pkg/service/utgen/assets/test_generation.toml [hunk 8]
```diff
@@ -109,65 +128,70 @@ existing_test_function_signature: |
...
 - test_behavior: |
     Test that the function returns the correct output for a single element list
-{{- if (or (eq .language "python") (eq .language "java")) }}
+{%- if language in ["python","java"] %}
   test_name: |
     test_single_element_list
-{{- else }}
+{%- else %}
   test_name: |
     ...
```
Comment (shivamsouravjha) [inline pkg/service/utgen/assets/test_generation.toml:136]: don't revert my changes https://github.com/keploy/keploy/pull/2406/files

File: pkg/service/utgen/gen.go [hunk 9]
```diff
@@ -235,13 +259,29 @@ func (g *UnitTestGenerator) Start(ctx context.Context) error {
...
 					return err
 				}
+
+				// if any test increases the coverage, set the flag to true
+				overallCovInc = overallCovInc || coverageInc
 			}
 
+			// if any of the test couldn't increase the coverage, revert the source code
+			if !overallCovInc {
+				err := revertSourceCode(g.srcPath, originalSrcCode, codeModified)
```
Comment (shivamsouravjha) [inline pkg/service/utgen/gen.go:280]: instead of using this boolean can't we use the error the way we do for test file changes

**Review by gouravkrosx — APPROVED** [review]
LFG!

**Pull request merged**